crates/precompile/src/bls12_381/blst.rs

```diff
@@ -314,3 +336,311 @@ pub(super) fn pairing_check(pairs: &[(blst_p1_affine, blst_p2_affine)]) -> bool
+
+    if subgroup_check {
+        // NB: Subgroup checks
+        //
+        // Scalar multiplications, MSMs and pairings MUST perform a subgroup check.
+        //
+        // Implementations SHOULD use the optimized subgroup check method:
+        //
+        // https://eips.ethereum.org/assets/eip-2537/fast_subgroup_checks
+        //
+        // On any input that fail the subgroup check, the precompile MUST return an error.
+        //
+        // As endomorphism acceleration requires input on the correct subgroup, implementers MAY
+        // use endomorphism acceleration.
+        if unsafe { !blst_p2_affine_in_g2(&out) } {
+            return Err(PrecompileError::Other("Element not in G2".to_string()));
+        }
+    }
+    Ok(out)
+}
+
+/// Checks whether or not the input represents a canonical field element
+/// returning the field element if successful.
+///
+/// Note: The field element is expected to be in big endian format.
+pub(super) fn read_fp(input: &[u8; FP_LENGTH]) -> Result<blst_fp, PrecompileError> {
+    if !is_valid_be(input) {
+        return Err(PrecompileError::Other("non-canonical fp value".to_string()));
+    }
+    let mut fp = blst_fp::default();
+    // SAFETY: `input` has fixed length, and `fp` is a blst value.
+    unsafe {
+        // This performs the check for canonical field elements
+        blst_fp_from_bendian(&mut fp, input.as_ptr());
+    }
+
+    Ok(fp)
+}
+
+/// Extracts a scalar from a 32 byte slice representation, decoding the input as a Big Endian
+/// unsigned integer. If the input is not exactly 32 bytes long, an error is returned.
+///
+/// From [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537):
+/// * A scalar for the multiplication operation is encoded as 32 bytes by performing BigEndian
+///   encoding of the corresponding (unsigned) integer.
+///
+/// We do not check that the scalar is a canonical Fr element, because the EIP specifies:
+/// * The corresponding integer is not required to be less than or equal than main subgroup order
+///   `q`.
+pub(super) fn read_scalar(input: &[u8]) -> Result<blst_scalar, PrecompileError> {
+    if input.len() != SCALAR_LENGTH {
+        return Err(PrecompileError::Other(format!(
+            "Input should be {SCALAR_LENGTH} bytes, was {}",
+            input.len()
+        )));
+    }
+
+    let mut out = blst_scalar::default();
+    // SAFETY: `input` length is checked previously, out is a blst value.
+    unsafe {
+        // Note: We do not use `blst_scalar_fr_check` here because, from EIP-2537:
+        //
+        // * The corresponding integer is not required to be less than or equal than main subgroup
+        // order `q`.
+        blst_scalar_from_bendian(&mut out, input.as_ptr())
+    };
+
+    Ok(out)
+}
+
+/// Checks if the input is a valid big-endian representation of a field element.
+fn is_valid_be(input: &[u8; 48]) -> bool {
+    for (i, modulo) in input.iter().zip(MODULUS_REPR.iter()) {
```

Comment:
can likely move this MODULUS_REPR constant into blst since it is not needed by any other backend

Reply:
I can do this in a separate PR since its a breaking change, as MODULUS_REPR is public

Reply:
ack